### PR TITLE
fix(47391): JSDoc/TSDoc @link with custom display text incorrectly shows vertical bar

### LINKED ISSUE #47391: JSDoc/TSDoc @link with custom display text incorrectly shows vertical bar
# Bug Report

### 🔎 Search Terms

jsdoc, tsdoc, link, display text, vertical bar, ...

### 🕗 Version & Regression Information
- Reproduced this successfully in VS Code with TypeScript v4.3.5, v4.4.4 and v4.5.4
- Not reproducible within the TypeScript Playground as it doesn't pretty-format `{@link ...}` tags

### 💻 Code and reproduction steps
Open VS Code, open an empty (or existing) file with "TypeScript" as language and insert the following code:
```ts
const A = 123;

/**
 * See {@link A | constant A} instead
 */
const B = 456;
```
Then hover over `B` to display the parsed documentation of that symbol.

### 🙁 Actual behavior

Hovering over `B` displays the vertical bar (`|`) in the link text:
Resulting tooltip

### 🙂 Expected behavior

Hovering over `B` should only display the text after the vertical bar:
image

### ℹ Extra information
Originally reported in [#46734](https://github.com/microsoft/TypeScript/issues/46734#issuecomment-1000470395) but told  to make a separate issue for it.

According to the [TSDoc specification](https://tsdoc.org/pages/tags/link/) and its dependant [API Extractor](https://api-extractor.com/pages/tsdoc/tag_link/) (where I stumbled upon this issue), if we want the display text to be different from the reference, we need to use `{@link DECLARATION_REFERENCE | DISPLAY_TEXT}`. But doing it this way results in the vertical bar being displayed in the link's display text.

While removing the vertical bar does make it display the correct text in the tooltip, this goes against the above documentation, their example code and even the official [tsdoc library](https://github.com/microsoft/tsdoc), as the latter will output (in my case) a `tsdoc-reference-missing-dot` warning, since it thinks it's still parsing a declaration reference due to not encountering a vertical bar.

While "testing" the above code in the Playground doesn't reproduce this issue (as it just renders `{@link A| constant A}` instead of link-ifying it), turning on the WIP AST Viewer under Settings (or use the one under Export if you change the `Tree mode`) shows that the `JSDocLink` has `| constant A` as `text`, which might already be wrong there if the trimming is supposed to happen earlier.

It's worth mentioning that [JSDoc does allow omitting the vertical bar](https://jsdoc.app/tags-inline-link.html), but it also supports having the vertical bar. Their examples showcase that if a vertical bar is used, it isn't displayed, similar to how TSDoc handles the vertical bar.

As far as I know this is a TypeScript issue, since the intellisense in VS Code should be provided by the TS language server and not VS Code itself? Perhaps I'm wrong about that.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #47391

### CODE REVIEW

**Comment by a-tarasyuk** [comment]
@sandersn Oke

File: tests/baselines/reference/quickInfoLink5.baseline [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          "text": "See ",
+          "kind": "text"
+        },
+        {
+          "text": "{@link ",
+          "kind": "link"
+        },
+        {
+          "text": "}",
+          "kind": "link"
```
Comment (sandersn) [inline tests/baselines/reference/quickInfoLink5.baseline:52]: this is an OK result but I'd prefer not to change the formatting of a malformed or unresolved link tag.

File: tests/cases/fourslash/quickInfoLink4.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/// <reference path="fourslash.ts" />
+
+////const A = 123;
+/////**
+//// * See {@link A | constant A} instead
```
Comment (sandersn) [inline tests/cases/fourslash/quickInfoLink4.ts:5]: looking at the code, it looks like `"| "`, `" |"` and `" | "` will all get skipped. Can you test all 3?

**Review by sandersn — CHANGES_REQUESTED** [review]
Looks good, I'd like a couple more test cases.

File: tests/baselines/reference/quickInfoLink5.baseline (hunk 1 above)
Comment (a-tarasyuk) [reply tests/baselines/reference/quickInfoLink5.baseline:52]: Do you mean the result should contain empty `linkText` for this case?
```json
{
    "text": "",
    "kind": "linkText"
},
```

**Review by sandersn — APPROVED** [review]

**Pull request merged**